File: examples/java-gradle/src/main/java/com/example/models/Product.java
package com.example.models;

import com.example.interfaces.Validator;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

/**
 * Product entity record.
 */
public record Product(Long id, String name, BigDecimal price, boolean inStock) implements Validator {

    public Product(String name, BigDecimal price) {
        this(null, name, price, true);
    }

    public Product(String name, double price) {
        this(null, name, BigDecimal.valueOf(price), true);
    }

    public Product withId(Long id) {
        return new Product(id, name, price, inStock);
    }

    public Product withStock(boolean inStock) {
        return new Product(id, name, price, inStock);
    }

    public Product applyDiscount(double percent) {
        BigDecimal discount = price.multiply(BigDecimal.valueOf(percent / 100));
        BigDecimal newPrice = price.subtract(discount).setScale(2, RoundingMode.HALF_UP);
        return new Product(id, name, newPrice, inStock);
    }

    public boolean isAvailable() {
        return inStock && price.compareTo(BigDecimal.ZERO) > 0;
    }

    @Override
    public List<String> validate() {
        List<String> errors = new ArrayList<>();

        if (name == null || name.isBlank()) {
            errors.add("Name cannot be empty");
        }

        if (price == null || price.compareTo(BigDecimal.ZERO) < 0) {
            errors.add("Price cannot be negative");
        }

        return errors;
    }

    @Override
    public boolean isValid() {
        return validate().isEmpty();
    }

    public String toJson() {
        return String.format(
                "{\"id\":%s,\"name\":\"%s\",\"price\":%s,\"inStock\":%s}",
                id, name, price, inStock
        );
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/models/Product.kt
package com.example.models

import com.example.interfaces.JsonSerializable
import com.example.interfaces.Validator
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Product data class.
 */
@Serializable
data class Product(
    val id: Long? = null,
    val name: String,
    val price: Double,
    val inStock: Boolean = true
) : Validator, JsonSerializable {

    val isAvailable: Boolean get() = inStock && price > 0

    fun withId(id: Long) = copy(id = id)

    fun withStock(inStock: Boolean) = copy(inStock = inStock)

    fun applyDiscount(percent: Double): Product {
        val newPrice = price * (1 - percent / 100)
        return copy(price = BigDecimal(newPrice).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    override fun validate(): List<String> = buildList {
        if (name.isBlank()) add("Name cannot be empty")
        if (price < 0) add("Price cannot be negative")
    }

    override fun toJson(): String = Json.encodeToString(this)

    override fun toPrettyJson(): String = Json { prettyPrint = true }.encodeToString(this)
}

/**
 * Extension function to calculate total price.
 */
fun List<Product>.totalPrice(): Double = sumOf { it.price }

/**
 * Type alias for product list.
 */
typealias ProductList = List<Product>
